lib/vrl/compiler/src/expression/literal.rs

```diff
@@ -51,7 +51,10 @@ impl TryFrom<Node<ast::Literal>> for Literal {
         let (span, lit) = node.take();
 
         let literal = match lit {
-            String(v) => Literal::String(Bytes::from(v)),
+            // Template strings have to be rewritten and compiled as an
+            // expression. Hence this variant should never be invoked.
+            String(_) => unreachable!(),
```

Comment:
This seems like an undesirable design. If template strings are not literals (they can be, if they only reference other literals), then they shouldn't be exposed as a literal expression.

Put another way, can we come up with a design that doesn't require this `unreachable`?

Reply:
I have removed the  `TryFrom` implementation, and have moved that functionality into the  `compiler::compile_literal` function. 

Now we can compile the string and have all the context required to do so.

Have also updated it to just return a `Literal::String` rather than rewrite if the template is just a single literal string.